src/Forms/GridField/GridFieldDataColumns.php

```diff
@@ -265,12 +292,21 @@ protected function castValue($gridField, $fieldName, $value)
             // If the value is an object, we do one of two things
             if (method_exists($value, 'Nice')) {
                 // If it has a "Nice" method, call that & make sure the result is safe
-                $value = nl2br(Convert::raw2xml($value->Nice()) ?? '');
+                $value = $value->Nice();
+                if ($this->getEscapeFields()) {
+                    $value = nl2br(Convert::raw2xml($value));
+                }
             } else {
-                // Otherwise call forTemplate - the result of this should already be safe
-                $value = $value->forTemplate();
+                $isHtml = is_a($value, DBHTMLText::class, false) || is_a($value, DBHTMLVarchar::class, false);
+                // For DBFields other than HTML variants, if we're not escaping values, get the raw value.
+                // Otherwise, check forTemplate() which is assumed to be safe.
+                if (!$isHtml && !$this->getEscapeFields() && is_a($value, DBField::class, false)) {
+                    $value = $value->RAW();
+                } else {
+                    $value = $value->forTemplate();
+                }
```

Comment:
```suggestion
                $isNonHtmlField = is_a($value, DBField::class, false) && (!is_a($value, DBHTMLText::class, false) && !is_a($value, DBHTMLVarchar::class, false));
                if ($isNonHtmlField && !$this->getEscapeFields()) {
                    // For DBFields other than HTML variants, if we're not escaping values, get the raw value.
                    $value = $value->RAW();
                } else {
                    // Otherwise, check forTemplate() which is assumed to be safe.
                    $value = $value->forTemplate();
                }
```

Reply:
`isNonHtmlField` implies to me that it's anything other than `DBHtmlText` or `DBHtmlVarchar`, which includes non `DBField`. But non `DBField` gets caught in the `else`.

I'll use your condition but name it `isDbFieldNonHtml` which makes it clear "this is a `DBField` - but not an HTML one".

Getting a bit convoluted here which is why I had the condition separated, but I don't think it's worth discussing so change made.

Reply:
Actually I've just removed the variable and dumped it all into the `if`. Also removed the unnecessary brackets (it's all `and` so no grouping is necessary)